clippy_lints/src/ptr.rs

```diff
@@ -111,6 +128,28 @@ fn check_fn(cx: &LateContext, decl: &FnDecl, fn_id: NodeId) {
             }
         }
     }
+
+    if let FunctionRetTy::Return(ref ty) = decl.output {
+        if let Some((out, MutMutable)) = get_rptr_lm(ty) {
+            if let Some(MutImmutable) = decl.inputs.iter()
+                     .filter_map(|ty| get_rptr_lm(ty))
+                     .filter(|&(lt, _)| lt.name == out.name)
+                     .fold(None, |x, (_, m)| match (x, m) {
+                        (Some(MutMutable), _) |
+                        (_, MutMutable) => Some(MutMutable),
+                        (_, m) => Some(m),
+                     }) {
+                span_lint(cx,
+                          MUT_FROM_REF,
+                          ty.span,
+                          "this function takes an immutable ref to return a mutable one:");
```

Comment:
Do we ever add a `:` at the end of an error message?

```rust
+error: this function takes an immutable ref to return a mutable one:
+  --> $DIR/mut_from_ref.rs:15:25
+   |
+15 |     fn ouch(x: &Foo) -> &mut Foo;
+   |                         ^^^^^^^^ 
```

this looks slightly weird because of the filename between the message and the code extract and because the span does not point to the function.

(let's all blame @oli-obk for converting to ui-tests and give me more reason to be pedantic :smile:)